src/target/riscv/riscv-013.c

```diff
@@ -2100,7 +2100,10 @@ static int assert_reset(struct target *target)
 
 	uint32_t control_base = set_field(0, DM_DMCONTROL_DMACTIVE, 1);
 
-	if (target->rtos) {
+	if (target_has_event_action(target, TARGET_EVENT_RESET_ASSERT)) {
+		/* Run the user-supplied script if there is one. */
+		target_handle_event(target, TARGET_EVENT_RESET_ASSERT);
+	} else if (target->rtos) {
```

Comment:
This looks like if a user configures a reset-assert script, then none of the regular reset stuff happens. http://openocd.org/doc/html/CPU-Configuration.html#targetevents says reset-assert is "Issued as part of reset processing," not instead of.

I don't know how replacing reset usually happens. From the script it looks like you really do want to just do your own reset processing. There must be other targets where that's required.

The simplest solution I can think of is to add a riscv config option to disable the regular reset if a reset-assert handler exists. Do you have time to find out how reset on noncompliant targets is handled on different architectures?

Reply:
Replacing the default handler is consistent with what other architectures do. Most targets don't have a standard architectural reset mechanism and so by default use SRST if it's present. For those architectures, when a `reset-assert` handler exists they call it instead of (not in addition to) asserting SRST.

Furthermore, a couple architectures, Cortex-M and MIPS M4K, do have architectural reset mechanisms, and they too skip these mechanisms when the config specifies a `reset-assert` handler.

Cortex-M has the following block right at the beginning of its reset function, which checks for the handler, running it and exiting early if found:
https://github.com/riscv/riscv-openocd/blob/2c909f8faae052ccc3cec82ed0d98b043f871ce3/src/target/cortex_m.c#L1082-L1090

MIPS M4K's code structure is even more obvious, with an explicit three-way switch between using SRST, a `reset-assert` handler, or the default architectural reset sequence:
https://github.com/riscv/riscv-openocd/blob/2c909f8faae052ccc3cec82ed0d98b043f871ce3/src/target/mips_m4k.c#L342-L368

Notably, this behavior doesn't mean that you can't write a config which keeps the default sequence and just augments it: that's exactly what `reset-assert-pre` and `reset-assert-post` are for.

Reply:
Thanks. Sounds like the documentation is just confusing in this case.